### PR TITLE
[Flax] Stateless schedulers, fixes and refactors

### PR DESCRIPTION
**Pull request by skirsten**
Hi, here some fixes and improvements for the Flax schedulers. Let me know what you think!
Sorry that its a huge PR with a single commit :sweat_smile: 

### Refactor schedulers to be completely stateless (all the state is in the params)

- No more state in the scheduler class
- No more [Implicit transfers](https://jax.readthedocs.io/en/latest/transfer_guard.html)
- Extracted common state (common state can also be reused from other schedulers)
- The shape and dtypes of the state returned by `set_timestamp` is now final and wont be changed by `step` (this reduces the amount of jit misses if jitting the scheduler separately)

### Added dtype param to schedulers

Leave it at fp32 though unless you want to lose all details in the image.

### Fix copy paste error in `add_noise` function

- The `add_noise` function and thus img2img were not working in `DDIM` and `DPMSolverMultistep`
- Extracted common logic so it can't happen again

### Removed all jax conditionals to fix performance bottleneck

- Using `jax.lax.cond` and `jax.lax.switch` causes the CPU to have to wait for the `pred` (even when jitted) causing the GPU pipeline to stall (not enough kernels scheduled). More info [here](https://github.com/google/jax/discussions/13586).
- If you notice that `PNDM` and `DPMSolverMultistep` were slower than `DDIM`, this was the reason.
- Usually this is most noticeable on fast GPU + slow CPU combo or if running a splitkernel (separately jitted scheduler instead of the megakernel as in this repo).
- Evaluating all branches instead has no noticeable performance impact.

### Fixed small bugs and improvements
- Added `v_prediction` where it was missing
- Made `DDPM` jitable. Though I'm not sure sure if it works correctly.
- Fixed `DPMSolverMultistep` not being able to start in the middle of a schedule. This caused img2img not to work.
- Made `LMSDiscrete` run. Its not jitable and I always get back a black image though.
- Probably some other stuff that I forgot about

### ~Validation~ (outdated)

I messed up so Pytorch is fp16 and Flax is bf16

| name                 | Pytorch                                                                                                                                   | Flax v0.10.2                                                                                                                             | Flax this PR                                                                                                                                  |
| -------------------- | ----------------------------------------------------------------------------------------------------------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------- |
| `DDIM`               | torch_ddim_0 10 2                | flax_ddim_0 10 2                | flax_ddim_0 11 0 dev0                |
| `DPMSolverMultistep` | torch_dpmsolver_multistep_0 10 2 | flax_dpmsolver_multistep_0 10 2 | flax_dpmsolver_multistep_0 11 0 dev0 |
| `PNDM` | torch_pndm_0 10 2   | flax_pndm_0 10 2   | flax_pndm_0 11 0 dev0   |

### CODE REVIEW

**Comment by pcuenca** [comment]
Hi @skirsten, this looks amazing! I see you are tweaking stuff, let me know when you want a review :)

**Comment by skirsten** [comment]
Hi @pcuenca, It should be ready for review now :sweat_smile:

**Comment by pcuenca** [comment]
Awesome, will do this week!

File: src/diffusers/schedulers/scheduling_common_flax.py [hunk 1]
```diff
@@ -0,0 +1,106 @@
+# Copyright 2022 The HuggingFace Team. All rights reserved.
```
Comment (patrickvonplaten) [inline src/diffusers/schedulers/scheduling_common_flax.py:1]: Can we maybe add this to `scheduling_utils_flax.py` instead ? :-) We usually don't have `_common_` files in `src/diffusers`

**Review by patrickvonplaten — COMMENTED** [review]
This looks super nice! Thanks a lot for working on this @skirsten :-) 

@pcuenca it'd be amazing if you could give this a try on a TPU

File: src/diffusers/schedulers/scheduling_common_flax.py (hunk 1 above)
Comment (skirsten) [reply src/diffusers/schedulers/scheduling_common_flax.py]: Done. I moved it to `scheduling_utils_flax.py`

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 2]
```diff
@@ -315,37 +327,34 @@ def step_prk(
...
-            return state.replace(cur_model_output=state.cur_model_output + 1 / 3 * model_output), model_output
-
-        def remainder_3(state: PNDMSchedulerState, model_output: jnp.ndarray, ets_at: int):
-            model_output = state.cur_model_output + 1 / 6 * model_output
-            return state.replace(cur_model_output=jnp.zeros_like(state.cur_model_output)), model_output
+        model_output = jax.lax.select(
+            (state.counter % 4) != 3,
+            model_output,  # remainder 0, 1, 2
+            state.cur_model_output + 1 / 6 * model_output,  # remainder 3
+        )
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:334]: These changes are all much cleaner this way. Thanks a lot!

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 3]
```diff
@@ -315,37 +327,34 @@ def step_prk(
...
-            model_output,
-            state.counter // 4,
+        state = state.replace(
+            cur_model_output=jax.lax.select_n(
+                state.counter % 4,
+                state.cur_model_output + 1 / 6 * model_output,  # remainder 0
+                state.cur_model_output + 1 / 3 * model_output,  # remainder 1
+                state.cur_model_output + 1 / 3 * model_output,  # remainder 2
+                jnp.zeros_like(state.cur_model_output),  # remainder 3
+            ),
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:343]: Nice!

File: src/diffusers/schedulers/scheduling_pndm_flax.py [hunk 4]
```diff
@@ -374,18 +383,20 @@ def step_plms(
...
-                "for more information."
-            )
+        # NOTE: There is no way to check in the jitted runtime if the prk mode was ran before
+        # if not self.config.skip_prk_steps and state.counter < 3:
+        #     raise ValueError(
+        #         f"{self.__class__} can only be run AFTER scheduler has been run "
+        #         "in 'prk' mode for at least 12 iterations "
+        #         "See: https://github.com/huggingface/diffusers/blob/main/src/diffusers/pipelines/pipeline_pndm.py "
+        #         "for more information."
+        #     )
```
Comment (pcuenca) [inline src/diffusers/schedulers/scheduling_pndm_flax.py:399]: You are right, feel free to remove the commented code if you want (maybe just leave the note?)

```suggestion
        # NOTE: There is no way to check in the jitted runtime if the prk mode was ran before
```

**Review by pcuenca — APPROVED** [review]
This is great! Much clearer code (and more efficient, according to the previous comments).

I tested PNDM, DDPM, DPM Solver and LMS Discrete on TPU v3-8 and these were my results:

- DPM Solver produces identical results as the previous version.
- There are some minor visual differences in PNDM. I haven't found the reason: tried to force `dtype` to `float32` when computing the betas but it didn't make a difference.
- LMS does not work in either version.
- DDPM produces noise in the new version. It crashed in the previous one.

I think we should merge this as it's so much better. My approach would be:
- Deal with DDPM and LMS in a followup PR, as they didn't work anyway in the previous implementation.
- If we can easily find a reason for the minor discrepancies in PNDM, let's try to apply it. I already spent a couple hours and couldn't find it, so I wouldn't spend much more time.

What do you think?

**Comment by pcuenca** [comment]
Maybe @patil-suraj wants to take a quick look too.

**Comment by skirsten** [comment]
> There are some minor visual differences in PNDM. I haven't found the reason: tried to force dtype to float32 when computing the betas but it didn't make a difference.

Are you using v prediction? I added the code for v prediction (from pytorch) and as can be seen in the sample images it has a  small visual difference.
